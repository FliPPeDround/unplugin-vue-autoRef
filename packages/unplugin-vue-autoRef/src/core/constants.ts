export const DEFINE_REF_MACROS = [
  'ref',
  'computed',
  'shallowRef',
  'customRef',
  'toRef',
]
