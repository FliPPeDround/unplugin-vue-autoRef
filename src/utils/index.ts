import { compileScript, parse } from 'vue/compiler-sfc'
import type { SFCDescriptor, SFCScriptBlock } from 'vue/compiler-sfc'

export type _SFCScriptBlock = Omit<
  SFCScriptBlock,
  'scriptAst' | 'scriptSetupAst'
>

export type SFCCompiled = Omit<SFCDescriptor, 'script' | 'scriptSetup'> & {
  script?: _SFCScriptBlock | null
  scriptSetup?: _SFCScriptBlock | null
  scriptCompiled: SFCScriptBlock
  lang: string | undefined
  isSetup: boolean
}

export const parseSFC = (code: string, id: string): SFCCompiled => {
  const { descriptor } = parse(code, {
    filename: id,
  })
  const isSetup = !!descriptor.scriptSetup
  const lang = (descriptor.script || descriptor.scriptSetup)?.lang

  let scriptCompiled: SFCScriptBlock | undefined
  return {
    ...descriptor,
    lang,
    isSetup,
    get scriptCompiled() {
      if (scriptCompiled)
        return scriptCompiled
      return (scriptCompiled = compileScript(descriptor, {
        id,
      }))
    },
  }
}
