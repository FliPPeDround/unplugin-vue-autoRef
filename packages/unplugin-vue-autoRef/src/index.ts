import { createUnplugin } from 'unplugin'
import { createFilter } from '@rollup/pluginutils'
import type { FilterPattern } from '@rollup/pluginutils'
import { transformMacros } from './core'

export interface Options {
  include?: FilterPattern
  exclude?: FilterPattern
  /**
   * $ref alias
   * @default ref
   */
  refAlias?: string
}

export type OptionsResolved = Omit<Required<Options>, 'exclude'> & {
  exclude?: FilterPattern
}

function resolveOption(options: Options): OptionsResolved {
  return {
    include: [/\.vue$/, /\.[jt]s[x]$/, /\.setup\.[cm]?[jt]sx?/],
    refAlias: '',
    ...options,
  }
}

const name = 'unplugin-vue-autoRef'

export default createUnplugin((userOptions: Options = {}) => {
  const options = resolveOption(userOptions)
  const filter = createFilter(options.include, options.exclude)

  // eslint-disable-next-line no-console, no-octal-escape
  console.log(`${'\033'}[33m \n ✈️ [${name}] Reactivity transform`)
  return {
    name,
    enforce: 'pre',
    transformInclude(id) {
      return filter(id)
    },
    transform(code, id) {
      try {
        return transformMacros(code, id, options.refAlias).code
      }
      catch (err: unknown) {
        this.error(`${name} ${err}`)
      }
    },
  }
})
