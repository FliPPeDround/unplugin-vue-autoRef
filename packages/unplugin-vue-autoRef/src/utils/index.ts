import { parse } from '@vue/compiler-sfc'
import { parse as babelParse } from '@babel/parser'
import type { SFCDescriptor, SFCScriptBlock } from '@vue/compiler-sfc'
import type MagicString from 'magic-string'
import traverse from '@babel/traverse'
import { DEFINE_REF_MACROS } from '../core/constants'

export type _SFCScriptBlock = Omit<
  SFCScriptBlock,
  'scriptAst' | 'scriptSetupAst'
>

export type SFCCompiled = Omit<SFCDescriptor, 'script' | 'scriptSetup'> & {
  script?: _SFCScriptBlock | null
  scriptSetup?: _SFCScriptBlock | null
  lang: string | undefined
  isSetup: boolean
}

export const parseSFC = (code: string, id: string): SFCCompiled => {
  const { descriptor } = parse(code, {
    filename: id,
  })
  const isSetup = !!descriptor.scriptSetup
  const lang = (descriptor.script || descriptor.scriptSetup)?.lang

  return {
    ...descriptor,
    lang,
    isSetup,
  }
}

export const parseMacros = (code: string, s: MagicString, refAlias: string, offset = 0) => {
  const ast = babelParse(code, {
    sourceType: 'unambiguous',
  })
  traverse(ast, {
    CallExpression(path) {
      const calleeName = path.get('callee').toString()
      if (DEFINE_REF_MACROS.includes(calleeName)) {
        const { callee } = path.node
        s.overwrite(
          callee.start! + offset,
          callee.end! + offset,
          calleeName === refAlias ? '$ref' : `\$${calleeName}`,
        )
      }
    },
  })
}
