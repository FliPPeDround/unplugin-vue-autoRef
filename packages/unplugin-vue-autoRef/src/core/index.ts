import MagicString from 'magic-string'
import { parseMacros, parseSFC } from '../utils'
import { DEFINE_REF_MACROS } from './constants'

function transformMacros(code: string, id: string) {
  if (!DEFINE_REF_MACROS.map(macros => code.includes(macros)).some(Boolean))
    return { code }

  const s = new MagicString(code)

  if (id.endsWith('.vue')) {
    const sfc = parseSFC(code, id)
    if (!sfc.script && !sfc.scriptSetup)
      return { code }
    const loc = sfc.scriptSetup?.loc || sfc.script?.loc
    parseMacros(loc!.source, s, loc!.start.offset)
  }
  else {
    parseMacros(code, s)
  }

  return {
    code: s.toString(),
    get map() {
      return s.generateMap({ hires: true })
    },
  }
}

export {
  transformMacros,
}
