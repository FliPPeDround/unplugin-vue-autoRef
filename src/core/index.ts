import traverse from '@babel/traverse'
import parser from '@babel/parser'
import { MagicString } from '@vue/compiler-sfc'
import { parseSFC } from './../utils'
import { DEFINE_REF_MACROS } from './constants'

function transform(code: string, id: string) {
  if (!DEFINE_REF_MACROS.map(macros => code.includes(macros)).some(Boolean))
    return { code }

  const sfc = parseSFC(code, id)
  if (!sfc.script && !sfc.scriptSetup)
    return { code }

  const loc = sfc.scriptSetup?.loc || sfc.script?.loc

  const s = new MagicString(code)

  const ast = parser.parse(loc!.source, {
    sourceType: 'unambiguous',
  })

  traverse(ast, {
    CallExpression(path) {
      const calleeName = path.get('callee').toString()
      if (DEFINE_REF_MACROS.includes(calleeName)) {
        const { callee } = path.node
        s.overwrite(callee.start! + loc!.start.offset, callee.end! + loc!.start.offset, `\$${calleeName}`)
      }
    },
  })

  return {
    code: s.toString(),
    get map() {
      return s.generateMap({ hires: true })
    },
  }
}

export {
  transform,
}
