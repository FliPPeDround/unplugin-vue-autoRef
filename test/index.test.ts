import { describe, expect, test } from 'vitest'
import { transform } from '../src/core'

describe('fixtures', async () => {
  const files = import.meta.glob('./fixtures/*/*.{vue,js,ts}', {
    eager: true,
    as: 'raw',
  })

  for (const [id, code] of Object.entries(files)) {
    test(id.replace(/\\/g, '/'), async () => {
      const exec = () => transform(code, id, '$').code
      if (id.includes('error'))
        expect(exec).toThrowErrorMatchingSnapshot()

      else
        expect(exec()).toMatchSnapshot()
    })
  }
})
