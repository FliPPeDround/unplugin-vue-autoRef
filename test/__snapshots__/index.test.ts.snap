// Vitest Snapshot v1

exports[`fixtures > ./fixtures/basic/basic.vue 1`] = `
"<script>
export default {
  setup() {
    const state = $ref(0)
    const count = $ref('1')
    return {
      state,
      count,
    }
  },
}
</script>

<template>
  {{ state }}
</template>
"
`;

exports[`fixtures > ./fixtures/noMacros.vue 1`] = `
"asdasd
"
`;

exports[`fixtures > ./fixtures/setup/computed.vue 1`] = `
"<script setup lang=\\"ts\\">
import { computed, ref } from 'vue'
const count = $ref(1)
const dobleCount = $computed(() => count.value * 2)
</script>

<template>
  <h1>{{ count }}</h1>
  <h1>{{ dobleCount }}</h1>
</template>
"
`;

exports[`fixtures > ./fixtures/setup/noMacros.vue 1`] = `
"asdasd
"
`;

exports[`fixtures > ./fixtures/setup/onMouted.vue 1`] = `
"<script setup lang=\\"ts\\">
import { onMounted, ref } from 'vue'
onMounted(() => {
  const count = $ref(1)
})
</script>

<template>
  <h1>{{ count }}</h1>
  <h1>{{ dobleCount }}</h1>
</template>
"
`;

exports[`fixtures > ./fixtures/setup/setup.vue 1`] = `
"<script setup lang=\\"ts\\">
import { ref } from 'vue'
const count = $ref(0)
const count2 = $ref(1)
const count3 = $ref(3)
</script>

<template>
  <h1>{{ count }}</h1>
</template>
"
`;
